ts/src/profiler.ts

```diff
@@ -13,47 +13,339 @@
+      body: reqBody,
+      json: true,
+    };
+
+    const [body, response] = await this.request(options);
+    if (!hasHttpStatusCode(response)) {
+      throw new Error('Server response missing status information.');
+    }
+    if (isErrorResponseStatusCode(response.statusCode)) {
+      let message: number|string = response.statusCode;
+      // tslint:disable-next-line: no-any
+      if ((response as any).statusMessage) {
+        message = response.statusMessage;
+      }
+      throw new Error(message.toString());
+    }
+    if (!isRequestProfile(body)) {
+      throw new Error(`Profile not valid: ${body}`);
+    }
+    return body;
+  }
+
+  /**
+   * Collects a profile of the type specified by the profileType field of prof.
+   * If any problem is encountered, like a problem collecting or uploading the
+   * profile, an error will be logged at the error level, but otherwise ignored.
+   *
+   * Public to allow for testing.
+   */
+  async profileAndUpload(prof: RequestProfile): Promise<void> {
+    try {
+      prof = await this.profile(prof);
+      prof.labels = this.profileLabels;
+    } catch (err) {
+      this.logger.debug(`Error collecting profile: ${err}`);
+      return;
+    }
+    const options = {
+      method: 'PATCH',
+      uri: API + '/' + prof.name,
+      body: prof,
+      json: true,
+    };
+    try {
+      const [body, response] = await this.request(options);
+      if (!hasHttpStatusCode(response)) {
+        throw new Error(
+            'Error uploading profile: server response missing status information.');
+      }
+      if (isErrorResponseStatusCode(response.statusCode)) {
+        let message: number|string = response.statusCode;
+        // tslint:disable-next-line: no-any
+        if ((response as any).statusMessage) {
+          message = response.statusMessage;
+        }
+        this.logger.error(`Error uploading profile: ${message}`);
+      }
+    } catch (err) {
+      this.logger.debug(`Error uploading profile: ${err}`);
+    }
+  }
+
+  /**
+   * Collects a profile of the type specified by profileType field of prof.
+   * If any problem is encountered, for example the profileType is not
+   * recognized or profiling is disabled for the specified profileType, an
+   * error will be thrown.
+   *
+   * Public to allow for testing.
+   */
+  async profile(prof: RequestProfile): Promise<RequestProfile> {
+    switch (prof.profileType) {
+      case ProfileTypes.Wall:
+        return await this.writeTimeProfile(prof);
+      case ProfileTypes.Heap:
+        return this.writeHeapProfile(prof);
+      default:
+        throw new Error(`Unexpected profile type ${prof.profileType}.`);
+    }
+  }
+
+  /**
+   * Collects a time profile, converts profile to compressed, base64 encoded
+   * string, and puts this string in profileBytes field of prof.
+   *
+   * Public to allow for testing.
+   */
+  async writeTimeProfile(prof: RequestProfile): Promise<RequestProfile> {
+    if (this.timeProfiler) {
```

Comment:
nit: reverse this if so that error is of the way early.

Reply:
Done.